File: correct/correct.py
import tifffile as tiff
import h5py
from libcorrect import correct_video_cython


def correct_video(in_file, motion_file, correction_file='',
                  first_frame=0, normalize=True,
                  motion_search_level=2, motion_search_size=5,
                  motion_patch_size=10, motion_patch_offset=7,
                  motion_x_range=1.0, motion_y_range=1.0,
                  shading_period=1000,
                  use_gpu=True, num_frames_per_batch=1000, num_threads=0):
    """
    Correct motion and shading in a video.

    Parameters
    ----------
    in_file : string or pathlib.Path
        Input file path of a multi-page tiff containig voltage imaging video.
    motion_file : string or pathlib.Path
        Output hdf5 file path to which estimated motion vectors will be saved.
    correction_file : string or pathlib.Path, optional
        Output tiff file path to which motion/shading-corrected video will be
        saved. The default is an empty string, in which case the video will be
        returned rather than saved.
    first_frame : integer, optional
        First frame number of the video. If nonzero, the input video will be
        treated as if it starts with that frame number by skipping the first
        frames. The default is zero.
    normalize : boolean, optional
        If True (default), the video intensity will be normalized.
    motion_search_level : integer, optional
        Max level of multiresolution motion correction. The default is 2.
        The GPU implementation does not support more than 2.
    motion_search_size : integer, optional
        [-search_size, +search_size] pixels will be searched in X and Y at
        each level as motion vector candidates. The default is 5.
    motion_patch_size : integer, optional
        [-patch_size, +patch_size] pixels will be used as a patch in X and Y
        when computing patch-based correlation in motion correction.
        The default is 10.
    motion_patch_offset : integer, optional
        Offset (both in X and Y) between adjacent patches. A larger value
        leads to fewer patches with smaller overlaps. The default is 7.
    motion_x_range : float, optional
        Fractional image range in X to be used in motion correction.
        For instance, 0.5 means patch-based correlation will be computed for
        50% of the image width around the center. The image outside of that
        range will not be referenced. This is useful when the image periphery
        does not contain much usable information to match.
        The default is 1.0 (full width).
    motion_y_range : float, optional
        Fractional image range in Y to be used in motion correction.
        The default is 1.0 (full height).
    shading_period : integer, optional
        Time period (in frames) for modeling shading variation.
        The default is 1000.
    use_gpu : boolean, optional
        GPUs will be used for the part of the algorithms whose GPU
        implementation is available. The default is True.
    num_frames_per_batch : integer, optional
        If use_gpu=True, the input video will be split into batches each having
        this number of frames, and will be processed one batch at a time.
        This will reduce the GPU memory usage and can improve performance as
        GPU data transfer can be overlapped with GPU processing.
        The default is 1000. This parameter will be ignored if use_gpu=False.
    num_threads : integer, optional
        The number of threads to run the preprocessing on. The default is 0,
        in which case all the available cores will be used.

    Returns
    -------
    c : 3D numpy.ndarray of float32
        Motion/shading-corrected video.

    """
    in_video = tiff.imread(in_file).astype('float32')
    in_video = in_video[first_frame:] # skip first frames
    c, x, y = correct_video_cython(in_video,
                                   1 if normalize else 0,
                                   motion_search_level, motion_search_size,
                                   motion_patch_size, motion_patch_offset,
                                   motion_x_range, motion_y_range,
                                   shading_period,
                                   1 if use_gpu else 0,
                                   num_frames_per_batch, num_threads)

    with h5py.File(motion_file, 'w') as f:
        f.create_dataset('x', data=x)
        f.create_dataset('y', data=y)

    if(correction_file):
        tiff.imwrite(correction_file, c, photometric='minisblack')
        return None
    else:
        return c

File: correct/lib/correct.cpp
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <omp.h>
#include <vector>

#include "malloc_util.h"
#include "image_util.h"
#include "TimerUtil.h"
#include "motion.h"
#include "shading.h"


static void normalize_intensity(size_t num_frames, size_t width, size_t height, float *img)
{
    const size_t num_pixels = width * height;

    // equalize average frame intensities
    #pragma omp parallel for
    for(size_t t = 0; t < num_frames; t++)
    {
        float sum = 0;
        for(size_t i = 0; i < num_pixels; i++)
        {
            sum += img[t * num_pixels + i];
        }
        const float scale = num_pixels / sum;
        for(size_t i = 0; i < num_pixels; i++)
        {
            img[t * num_pixels + i] *= scale;
        }
    }

    // rescale intensities so that [min, max] = [0, 1]
    float min = img[0];
    float max = img[0];
    #pragma omp parallel for reduction(min: min) reduction(max: max)
    for(size_t t = 0; t < num_frames; t++)
    for(size_t i = 0; i < num_pixels; i++)
    {
        float val = img[t * num_pixels + i];
        if(min > val) min = val;
        if(max < val) max = val;
    }

    const float scale = 1.0 / (max - min);
    #pragma omp parallel for
    for(size_t t = 0; t < num_frames; t++)
    for(size_t i = 0; i < num_pixels; i++)
    {
        img[t * num_pixels + i] = (img[t * num_pixels + i] - min) * scale;
    }
}


void correct_video_cpp(int num_frames, int height, int width,
                       float *in_image,
                       float **out_image,
                       float **out_x, float **out_y,
                       int normalize,
                       int motion_search_level, int motion_search_size,
                       int motion_patch_size, int motion_patch_offset,
                       float motion_x_range, float motion_y_range,
                       int shading_period,
                       int use_gpu, int num_frames_per_batch, int num_threads)
{
    if(num_threads > 0)
    {
        omp_set_num_threads(num_threads);
    }
    
    const size_t t = num_frames;
    const size_t h = height;
    const size_t w = width;
    
    motion_param_t motion_param;
    motion_param.level = motion_search_level;
    motion_param.search_size = motion_search_size;
    motion_param.patch_size = motion_patch_size;
    motion_param.patch_offset = motion_patch_offset;
    motion_param.x_range = motion_x_range;
    motion_param.y_range = motion_y_range;
    //motion_param.thresh_xy = 10.0;
    motion_param.length = 1000;
    motion_param.thresh_c = 1.0;

    shading_param_t shading_param;
    shading_param.period = shading_period;

    
    TimerUtil *tu;

    if(normalize)
    {
        tu = new TimerUtil("intensity normalization");
        normalize_intensity(t, w, h, in_image);
        delete tu;
    }

    std::vector<motion_t> motion_list;
    *out_image = malloc_float1d(t * h * w);

    tu = new TimerUtil("motion correction");
    if(use_gpu)
    {
        motion_list = correct_motion_gpu(motion_param, t, w, h,
                                         num_frames_per_batch,
                                         in_image, *out_image);
    }
    else
    {
        float ***tmp = malloc_float3d(t, w, h);
        copy1d_to_3d(t, w, h, in_image, tmp);
        motion_list = correct_motion(motion_param, t, w, h, tmp);
        copy3d_to_1d(t, w, h, tmp, *out_image);
        free_float3d(tmp);
    }
    delete tu;

    float min_x = 0, max_x = 0, min_y = 0, max_y = 0;
    *out_x = malloc_float1d(t);
    *out_y = malloc_float1d(t);
    float *px = *out_x;
    float *py = *out_y;
    #pragma omp parallel for reduction(min: min_x, min_y) reduction(max: max_x, max_y)
    for(int k = 0; k < t; k++)
    {
        float x = motion_list[k].x;
        float y = motion_list[k].y;
        px[k] = x;
        py[k] = y;
        if(min_x > x) min_x = x;
        if(max_x < x) max_x = x;
        if(min_y > y) min_y = y;
        if(max_y < y) max_y = y;
    }
    printf("(x, y) in [%.1f, %.1f] x [%.1f, %.1f]\n", min_x, max_x, min_y, max_y);

    tu = new TimerUtil("shading correction");
    correct_shading(shading_param, t, w, h, *out_image, motion_list);
    delete tu;
}

